### PR TITLE
make needless_return work with void functions

### PR DESCRIPTION
**Pull request by d-dorazio**
fixes https://github.com/rust-lang/rust-clippy/issues/4181.

changelog: make needless_return work with void functions.

I don't think the failure is related to my changes, but I'm not sure :thinking:

### CODE REVIEW

**Comment by phansch** [comment]
Closing & Re-opening to trigger a new build

File: clippy_lints/src/returns.rs [hunk 1]
```diff
@@ -117,32 +123,61 @@ impl Return {
...
+                        span_lint_and_then(cx, NEEDLESS_RETURN, ret_span, "unneeded return statement", |db| {
+                            db.span_suggestion(
+                                ret_span,
+                                "remove `return`",
+                                String::new(),
+                                Applicability::MachineApplicable,
+                            );
+                        });
+                    }
+                    RetReplacement::Unit => {
```
Comment (flip1995) [inline clippy_lints/src/returns.rs]: Can you add a test for this case?

File: clippy_lints/src/returns.rs [hunk 2]
```diff
@@ -117,32 +123,61 @@ impl Return {
...
+                        span_lint_and_then(cx, NEEDLESS_RETURN, ret_span, "unneeded return statement", |db| {
+                            db.span_suggestion(
+                                ret_span,
+                                "replace `return` with the unit type `()`",
```
Comment (flip1995) [inline clippy_lints/src/returns.rs]: The test case for this revealed a NIT:

```suggestion
                                "replace `return` with the unit type",
```

File: clippy_lints/src/returns.rs [hunk 3]
```diff
@@ -117,32 +123,61 @@ impl Return {
...
+            Some(inner_span) => {
+                if in_external_macro(cx.sess(), inner_span) || in_macro_or_desugar(inner_span) {
+                    return;
+                }
+
+                span_lint_and_then(cx, NEEDLESS_RETURN, ret_span, "unneeded return statement", |db| {
+                    if let Some(snippet) = snippet_opt(cx, inner_span) {
+                        db.span_suggestion(
+                            ret_span,
+                            "remove `return` as shown",
```
Comment (flip1995) [inline clippy_lints/src/returns.rs]: See https://doc.rust-lang.org/nightly/nightly-rustc/syntax/diagnostics/plugin/struct.Diagnostic.html#method.span_suggestion:

```suggestion
                            "remove `return`",
```

Comment (d-dorazio) [reply clippy_lints/src/returns.rs]: gotcha! we should be good now.

**Comment by flip1995** [comment]
Thanks!

@bors r+

**Comment by flip1995** [comment]
waiting for #4227

**Pull request merged**